_posts/2023-10-02-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,457 @@
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.clients.consumer.OffsetOutOfRangeException;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.RebalanceInProgressException;
+import org.apache.kafka.common.errors.RetriableException;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.LongDeserializer;
+
+import static java.time.Duration.ofMillis;
+import static java.util.Collections.singleton;
+
+public class Consumer implements ConsumerRebalanceListener, OffsetCommitCallback {
+    // Constants for configuration
+    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
+    private static final String GROUP_ID = "my-group";
+    private static final long POLL_TIMEOUT_MS = 1_000L;
+    private static final String TOPIC_NAME = "my-topic";
+    private static final long NUM_MESSAGES = 50;
+    private static final long PROCESSING_DELAY_MS = 1000L;
+
+    private KafkaConsumer<Long, byte[]> kafkaConsumer;
+    protected AtomicLong messageCount = new AtomicLong(0);
+    private Map<TopicPartition, OffsetAndMetadata> pendingOffsets = new HashMap<>();
+
+    public static void main(String[] args) {
+        new Consumer().run();
+    }
+
+    public void run() {
+        System.out.println("Running consumer");
+
+        // Create a Kafka consumer instance
+        // This consumer receives messages from the Kafka topic asynchronously
+        try (var consumer = createKafkaConsumer()) {
+            kafkaConsumer = consumer;
+            consumer.subscribe(singleton(TOPIC_NAME), this);
+            System.out.printf("Subscribed to %s%n", TOPIC_NAME);
+            while (messageCount.get() < NUM_MESSAGES) {
+                try {
+                    // Poll for new records from Kafka
+                    ConsumerRecords<Long, byte[]> records = consumer.poll(ofMillis(POLL_TIMEOUT_MS));
+                    if (!records.isEmpty()) {
+                        for (ConsumerRecord<Long, byte[]> record : records) {
+                            System.out.printf("Record fetched from %s-%d with offset %d%n",
+                                record.topic(), record.partition(), record.offset());
+                            sleep(PROCESSING_DELAY_MS);
+
+                            // Track pending offsets for commit
+                            pendingOffsets.put(new TopicPartition(record.topic(), record.partition()),
+                                new OffsetAndMetadata(record.offset() + 1, null));
+                            if (messageCount.incrementAndGet() == NUM_MESSAGES) {
+                                break;
+                            }
+                        }
+                        // Commit pending offsets asynchronously
+                        consumer.commitAsync(pendingOffsets, this);
```

Comment:
the way you are using the `commitAsync` let me think you don't need to gather the pending offsets but just calling the `commitAsync(this)`.
I mean, you are calling it outside of the for loop so when all records are processed. So what you expect is to commit offset related to the last processed record (with higher offset) to mark the final point the consumer reached on that partition. This is what actually the `commitAsync(this)` does without providing a list of offsets.
Using a list of pending offsets would be useful if you are committing offsets more frequently after polling more records (i.e. poll 10 records but you want to commit offset every 3-4 records).
I am interested to know what the other thinks about it or you used gathering the pending offsets for a specific reason.

Reply:
Left unchanged

Reply:
Ok fine with me.